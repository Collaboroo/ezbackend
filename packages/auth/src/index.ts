import 'fastify-passport';
export * from './auth';
export * from './ez-user';
export * from './providers';
