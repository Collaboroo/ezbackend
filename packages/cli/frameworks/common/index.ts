import { EzModel, EzRouter } from "@ezbackend/common";
import { DataTypes } from "sequelize";
import _ from "lodash";

export const sample = new EzModel("sample", {
  string: {
    type: DataTypes.STRING,
    allowNull: false,
  },
  int: {
    type: DataTypes.INTEGER,
  },
  float: {
    type: DataTypes.FLOAT,
  },
  double: {
    type: DataTypes.DOUBLE
  },
  real: {
    type: DataTypes.REAL
  },
  enum: {
    type: DataTypes.ENUM,
    values: ['one','two','three']
  },
  data: {
    type: DataTypes.DATE
  },
  array: {
    type: DataTypes.ARRAY(DataTypes.TEXT)
  },
  uuid: {
    type: DataTypes.UUID
  },
  json: {
    type: DataTypes.JSON
  }
});

export const user = new EzModel("user", {
  name: {
    type: DataTypes.STRING,
    allowNull: false,
  },
});

export const program = new EzModel("program", {});

program.hasOne(user);

// export const customEndpoint = new EzRouter("custom")

// customEndpoint.registerRoute({
//   method: 'GET',
//   url: '/',
//   schema: {
//     body: {

//     },
//     response: {
//       200: {}
//     }
//   },
//   handler(req,res) {
//     res.send('hi')
//   }
// })

  
// export const customEndpoint = new EzRouter("custom")

// customEndpoint.registerRoute({
//   method: 'GET',
//   url: '/',
//   schema: {
//     body: {

//     },
//     response: {
//       200: {}
//     }
//   },
//   handler(req,res) {
//     res.send('hi')
//   }
// })