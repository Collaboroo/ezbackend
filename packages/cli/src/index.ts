import { Command } from "commander";
import { sync as readPackageUpSync } from "read-pkg-up";
import initiate from "./initiate";

//URGENT TODO: Make dependencies core dependencies not dev dependencies during installation
//URGENT TODO: Add @ezbackend/core to dependencies

const pkg = readPackageUpSync({ cwd: __dirname })?.packageJson;

const program = new Command();

program
  .command("init")
  .description("Initialize ezbackend into your project.")
  .option("-f --force", "Force add ezbackend")
  .action((options) => initiate(options, pkg));

program.parse(process.argv);

