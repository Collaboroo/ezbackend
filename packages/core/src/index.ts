type IEzbPlugin = () => void;

export type IEzbPlugins = {
  preInit: Array<IEzbPlugin>;
  init: IEzbPlugin | null;
  postInit: Array<IEzbPlugin>;
  preHandler: Array<IEzbPlugin>;
  handler: IEzbPlugin | null;
  postHandler: Array<IEzbPlugin>;
  preRun: Array<IEzbPlugin>;
  run: IEzbPlugin | null;
  postRun: Array<IEzbPlugin>;
};

export class EzBackend {
  plugins: IEzbPlugins;

  private static instance: EzBackend;

  constructor() {
    this.plugins = {
      preInit: [],
      init: () => {},
      postInit: [],
      preHandler: [],
      handler: () => {},
      postHandler: [],
      preRun: [],
      run: () => {},
      postRun: [],
    };
  }

  public static initializeApp() {
    if (!EzBackend.instance) {
      EzBackend.instance = new EzBackend();
    }
  }

  public static app(): EzBackend {
    EzBackend.initializeApp();
    return EzBackend.instance;
  }

  public static start(): void {
    const ezb = EzBackend.app();
    const plugins = ezb.plugins;

    plugins.preInit.forEach((plugin) => {
      plugin();
    });

    plugins.init();

    plugins.postInit.forEach((plugin) => {
      plugin();
    });

    plugins.preHandler.forEach((plugin) => {
      plugin();
    });

    plugins.handler();

    plugins.postHandler.forEach((plugin) => {
      plugin();
    });

    plugins.preRun.forEach((plugin) => {
      plugin();
    });

    plugins.run();

    plugins.postRun.forEach((plugin) => {
      plugin();
    });
  }
}
