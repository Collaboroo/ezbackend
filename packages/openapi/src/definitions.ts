import {EzBackend} from '@ezbackend/core'
import {SwaggerOptions} from 'fastify-swagger'
import {FastifyRegisterOptions} from 'fastify'
import type {OpenAPIV3} from 'openapi-types'

declare module '@ezbackend/core' {
    interface EzBackend{
        openapi: {
            config: FastifyRegisterOptions<SwaggerOptions>,
        }
    }
  }
  