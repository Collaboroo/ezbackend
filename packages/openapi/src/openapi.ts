import { SwaggerOptions, fastifySwagger } from 'fastify-swagger'

import { EzApp } from "@ezbackend/common"
import { EzBackendOpts } from "@ezbackend/common"
import type { FastifyRegisterOptions } from 'fastify'
import { PluginScope } from "@ezbackend/core"

declare module '@ezbackend/common' {
    interface EzBackendOpts {
        openAPI: FastifyRegisterOptions<SwaggerOptions> | undefined
    }
}

const defaultConfig: EzBackendOpts['openAPI'] = {
    prefix: "/docs",
    routePrefix: "/docs",
    exposeRoute: true,
    //TODO: Figure out why its logging so much
    logLevel: 'warn',
    openapi: {
        info: {
            title: "EzBackend API",
            description: "Automatically generated documentation for EzBackend",
            version: "1.0.0",
        },
        externalDocs: {
            url: "https://github.com/kapydev/ezbackend",
            description: "Find more info here",
        }
    },
}

export class EzOpenAPI extends EzApp {
    constructor() {
        super()

        this.setDefaultOpts(defaultConfig)

        this.setHandler('Add Swagger Plugin', async (instance, fullOpts) => {

            const opts = this.getOpts('openAPI', fullOpts)

            instance.server.register(fastifySwagger,opts)
        })

        this.scope = PluginScope.PARENT
    }
}