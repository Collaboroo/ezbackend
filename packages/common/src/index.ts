import { IOptions } from "./definitions";
import { IEzbConfig } from "@ezbackend/core";
import { fastify } from "fastify";
import path from "path";
import { EzBackend } from "./definitions";
import { mixedInstance } from "avvio";
import { createConnection } from "typeorm";
import { RouteOptions } from "fastify";
import { APIGenerator } from "./models";
import {convert} from './models/typeorm-json-schema'

const ezb = EzBackend.app();

//Configure defaults
ezb.plugins.init = async (
  ezb: mixedInstance<EzBackend>,
  opts: IEzbConfig & IOptions,
  cb
) => {
  ezb.orm = await createConnection(opts.orm);
  ezb.server = fastify(opts.server);
  cb();
};

ezb.plugins.handler = (ezb: mixedInstance<EzBackend>, opts: IEzbConfig, cb) => {
  const customEzbPath =
    opts.entryPoint ?? path.join(process.cwd(), ".ezb/index.ts");
  require(customEzbPath);
  ezb.models.forEach((model) => {
    const repo = ezb.orm.getRepository(model);

    //LEFT OFF
    // const metaData = ezb.orm.getMetadata(model)
    // console.log(convert(metaData))

    const generator = new APIGenerator(repo, { prefix: repo.metadata.name });
    generator.generateRoutes();
  });
  cb();
};

ezb.plugins.run = async (
  ezb: mixedInstance<EzBackend>,
  opts: IEzbConfig,
  cb
) => {
  await ezb.server.listen(opts.port, function (err, address) {
    if (err) {
      console.error(err);
      process.exit(1);
    }
  });
  cb();
};

export { EzBackend } from "./definitions";
export { EzModel, response, APIGenerator } from "./models";
