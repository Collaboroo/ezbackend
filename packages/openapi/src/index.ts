import { IEzbConfig } from '@ezbackend/core';
import { mixedInstance } from 'avvio';
import { EzBackend } from '@ezbackend/core'
import { APIGenerator, getPrimaryColName } from "@ezbackend/common";
import { RouteOptions } from 'fastify';
import fastifySwagger from "fastify-swagger"
import chalk from 'chalk'
import './definitions'

const ezb = EzBackend.app();

ezb.openapi = {
  config: {
    prefix: "/docs",
    routePrefix: "/docs",
    exposeRoute: true,
    //TODO: Figure out why its logging so much
    logLevel: 'warn',
    openapi: {
      info: {
        title: "EzBackend API",
        description: "Automatically generated documentation for EzBackend",
        version: "1.0.0",
      },
      externalDocs: {
        url: "https://github.com/kapydev/ezbackend",
        description: "Find more info here",
      }
    },
    
  }
}

// components: {
//   securitySchemes: {
//     OAuth2: {
//       type: "oauth2",
//       description: "## ⚠️Do not fill client id, just click __'Authorize'__ [(explanation)](http://google.com)",
//       flows:
//       {
//         implicit: {
//           authorizationUrl: 'http://localhost:8888/User/auth/google/login',
//           scopes: {
//           }
//         }
//       }
//     }
//   }
// }
//   },
// }

//Configure defaults
ezb.plugins.postInit.push((ezb, opts, cb) => {
  ezb.server.register(fastifySwagger,ezb.openapi.config);
cb()
});

const originalGenerators = APIGenerator.getGenerators()

if (originalGenerators['createOne'] !== undefined) {
  const oldGenerator = originalGenerators['createOne']
  APIGenerator.setGenerator("createOne", (repo) => {
    const routeDetails = oldGenerator(repo)
    const generatedCols = repo.metadata.columns.filter(col => col.isGenerated).map(col => col.propertyName)
    return {
      ...routeDetails,
      schema: {
        ...routeDetails.schema,
        summary: `Create ${repo.metadata.name}`,
        tags: [repo.metadata.name],
        description: `During creation, you are not allowed to specify the values of generated columns (e.g. ${generatedCols.toString()}).
        All non nullable columns must be specified on creation`
      }
    }
  });
}

if (originalGenerators['getOne'] !== undefined) {
  const oldGenerator = originalGenerators['getOne']
  APIGenerator.setGenerator("getOne", (repo) => {
    const primaryColName = getPrimaryColName(repo)
    const routeDetails = oldGenerator(repo)
    return {
      ...routeDetails,
      schema: {
        ...routeDetails.schema,
        summary: `Get ${repo.metadata.name} by ${primaryColName}`,
        tags: [repo.metadata.name],
        description: `If the ${primaryColName} does not contain the value specified in the url paramters, there will be a 'not found' error.`
      }
    }
  });
}

if (originalGenerators['getAll'] !== undefined) {
  const oldGenerator = originalGenerators['getAll']
  APIGenerator.setGenerator("getAll", (repo) => {
    const routeDetails = oldGenerator(repo)
    return {
      ...routeDetails,
      schema: {
        ...routeDetails.schema,
        summary: `Get all ${repo.metadata.name} instances`,
        tags: [repo.metadata.name],
        description: `If none exist, an empty array is returned`
      }
    }
  });
}


if (originalGenerators['updateOne'] !== undefined) {
  const oldGenerator = originalGenerators['updateOne']
  APIGenerator.setGenerator("updateOne", (repo) => {
    const primaryColName = getPrimaryColName(repo)
    const routeDetails = oldGenerator(repo)
    const generatedCols = repo.metadata.columns.filter(col => col.isGenerated).map(col => col.propertyName)
    return {
      ...routeDetails,
      schema: {
        ...routeDetails.schema,
        summary: `Update ${repo.metadata.name} by ${primaryColName}`,
        tags: [repo.metadata.name],
        description: `The ${repo.metadata.name} with the ${primaryColName} specified must exist, otherwise a 'not found' error is returned
        During creation, you are not allowed to specify the values of generated columns (e.g. ${generatedCols.toString()})`
      }
    }
  });
}

if (originalGenerators['deleteOne'] !== undefined) {
  const oldGenerator = originalGenerators['deleteOne']
  APIGenerator.setGenerator("deleteOne", (repo) => {
    const primaryColName = getPrimaryColName(repo)
    const routeDetails = oldGenerator(repo)
    return {
      ...routeDetails,
      schema: {
        ...routeDetails.schema,
        summary: `Delete ${repo.metadata.name} by ${primaryColName}`,
        tags: [repo.metadata.name],
        description: `The ${repo.metadata.name} with the ${primaryColName} specified must exist, otherwise a 'not found' error is returned`
      }
    }
  });
}





//URGENT TODO: Generate Schemas for openapi properly
//TODO: Make page when user reopens swagger
ezb.plugins.postRun.push((ezb: mixedInstance<EzBackend>, opts: IEzbConfig, cb) => {
  // ezb.server.swagger();
  if (opts.port) {
    console.log(chalk.greenBright(`\nView your auto-generated Documentation at `) + chalk.yellow.underline(`http://localhost:${opts.port}/docs\n`))
  }
  cb()
})

export * from './definitions'