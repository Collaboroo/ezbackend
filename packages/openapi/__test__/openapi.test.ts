import { EzBackend, EzModel, Type } from "@ezbackend/common";

import { EzOpenAPI } from '../src'
import type { FastifyInstance } from 'fastify'

//TODO: Figure if there is a better way of getting this data
function getInternalInstance(ezb: EzBackend) {
    //@ts-ignore
    return ezb.instance._lastUsed.server
}

let app: EzBackend

const defaultConfig = {
    backend: {
        fastify: {
            logger: false
        },
        typeorm: {
            database: ':memory:'
        }
    }
}

beforeEach(() => {
    app = new EzBackend()

    //Prevent server from starting
    app.removeHook("_run", "Run Fastify Server")
})

afterEach(async () => {
    const instance = getInternalInstance(app)
    await instance.orm.close();
    await instance._server.close();
});

describe("Basic Usage", () => {
    it("Should be able to render the docs", async () => {
        app.addApp(new EzOpenAPI())

        await app.start(defaultConfig)

        const server: FastifyInstance = getInternalInstance(app)._server

        const response = await server.inject({
            method: "GET",
            url: "/docs/static/index.html",
        })

        expect(response.statusCode).toBe(200)
    })

    it("Should have schemas", async () => {

        app.addApp('openapi', new EzOpenAPI())

        const dummyModel = new EzModel('model', {
            var1: Type.VARCHAR,
            var2: Type.VARCHAR
        })

        app.addApp('model', dummyModel)

        await app.start(defaultConfig)

        const server: FastifyInstance = app.getInternalServer()

        const response2 = await server.inject({
            method: "GET",
            url: "/docs",
        })

        const response = await server.inject({
            method: "GET",
            url: "/docs/json",
        })


        expect(response.statusCode).toBe(200)
        expect(JSON.parse(response.body)['components']).toMatchSnapshot()
    })

})