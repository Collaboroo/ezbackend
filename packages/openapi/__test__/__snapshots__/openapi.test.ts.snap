// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic Usage Should have schemas 1`] = `
Object {
  "schemas": Object {
    "def-0": Object {
      "properties": Object {
        "error": Object {
          "type": "string",
        },
        "message": Object {
          "type": "string",
        },
        "statusCode": Object {
          "type": "number",
        },
      },
      "title": "ErrorResponse",
      "type": "object",
    },
    "def-1": Object {
      "properties": Object {
        "var1": Object {
          "type": "string",
        },
        "var2": Object {
          "type": "string",
        },
      },
      "required": Array [
        "var1",
        "var2",
      ],
      "title": "createSchema-model",
      "type": "object",
    },
    "def-2": Object {
      "properties": Object {
        "var1": Object {
          "type": "string",
        },
        "var2": Object {
          "type": "string",
        },
      },
      "required": Array [
        "var1",
        "var2",
      ],
      "title": "formCreateSchema-model",
      "type": "object",
    },
    "def-3": Object {
      "properties": Object {
        "var1": Object {
          "type": "string",
        },
        "var2": Object {
          "type": "string",
        },
      },
      "title": "formUpdateSchema-model",
      "type": "object",
    },
    "def-4": Object {
      "properties": Object {
        "var1": Object {
          "type": "string",
        },
        "var2": Object {
          "type": "string",
        },
      },
      "title": "updateSchema-model",
      "type": "object",
    },
    "def-5": Object {
      "properties": Object {
        "id": Object {
          "type": "number",
        },
        "var1": Object {
          "type": "string",
        },
        "var2": Object {
          "type": "string",
        },
      },
      "title": "fullSchema-model",
      "type": "object",
    },
  },
}
`;
